File: go/p4.go
package main

import "fmt"

func NumDigits(n int) int {
        r := 0
        switch {
        case n < 10: r = 1 
        case n < 100: r = 2 
        case n < 1000: r = 3 
        case n < 10000: r = 4 
        case n < 100000: r = 5 
        default: r = 6 
        }
        return r
}


func Pow10(n int) int {
        r := 0
        switch {
        case n == 0: r = 1
        case n == 1: r = 10
        case n == 2: r = 100
        case n == 3: r = 1000
        case n == 4: r = 10000
        case n == 5: r = 100000
        default: r = 1000000
        }
        return r
}


func NthDigit(n, d int) int {
        return (n % Pow10(d+1)) / Pow10(d);
}


func IsPalindrome(n int) bool {
        digcount := NumDigits(n) - 1
        for i, j := 0, digcount; i < j; i, j = i+1, j-1 {
                if NthDigit(n, i) != NthDigit(n, j) {
                        return false
                }
        }
        return true;
}
                        

func main() {
        max := 0
        for i := 100; i <= 999; i++ {
                for j := 100; j <= 999; j++ {
                        prod := i * j
                        if (IsPalindrome(prod) && prod > max) {
                                max = prod
                        }
                }
        }
        fmt.Println(max)
}

        

File: ocaml/p4.ml
open Core.Std

(* Ugly, but fast *)
let pow10 = function
    | 0 -> 1 | 1 -> 10 | 2 -> 100 | 3 -> 1000
    | 4 -> 10000 | 5 -> 100000 | _ -> 1000000

let count_digits = function
    | n when n < 10 -> 1
    | n when n < 100 -> 2
    | n when n < 1000 -> 3
    | n when n < 10000 -> 4
    | n when n < 100000 -> 5
    | _ -> 6

let digit d n = (n mod (pow10 (d + 1))) / (pow10 d)

let palindrome n =
    let count = (count_digits n) - 1 in
    let rec loop d =
        let d' = count - d in
        if d >= d' then true
        else if (digit d n) <> (digit d' n) then false
        else loop (d + 1)
    in
    loop 0

let rec largest_palindrome max i j =
    if i > 999 then max
    else
        let prod = i * j in
        let i', j' = if j = 999 then ((i+1), 100) else (i, (j+1)) in
        let max' = 
            if (palindrome prod) && (prod > max) then prod else max
        in
        largest_palindrome max' i' j'
    

let () = printf "%d\n" (largest_palindrome 0 100 100)
